staging/src/k8s.io/apiserver/pkg/registry/rest/validate.go

```diff
@@ -106,3 +111,137 @@ func parseSubresourcePath(subresourcePath string) ([]string, error) {
+		recommendation = "Consider disabling the DeclarativeValidationTakeover feature gate."
+	}
+
+	matcher := fldtest.ErrorMatcher{}.ByType().ByField().ByOrigin().RequireOriginWhenInvalid()
+	matchedDeclarative := make([]bool, len(declarativeErrs))
+	// Match each "covered" imperative error to a declarative error.
+	for _, iErr := range imperativeErrs {
+		if !iErr.CoveredByDeclarative {
+			continue
+		}
+		foundMatch := false
+		for j, dErr := range declarativeErrs {
+			if !matchedDeclarative[j] && matcher.Matches(iErr, dErr) {
+				matchedDeclarative[j] = true
+				foundMatch = true
+				break
+			}
+		}
+		if !foundMatch {
+			mismatchDetails = append(mismatchDetails,
+				fmt.Sprintf(
+					"Unexpected difference between hand written validation and declarative validation error results, unmatched error(s) found %s. "+
+						"This may indicate an issue with declarative validation. %s",
+					matcher.Render(iErr),
+					recommendation,
+				),
+			)
+		}
+	}
+	// Any remaining unmatched declarative errors are considered "extra".
+	for j, dErr := range declarativeErrs {
+		if !matchedDeclarative[j] {
+			mismatchDetails = append(mismatchDetails,
+				fmt.Sprintf(
+					"Unexpected difference between hand written validation and declarative validation error results, extra error(s) found %s. "+
+						"This may indicate an issue with declarative validation. %s",
+					matcher.Render(dErr),
+					recommendation,
+				),
+			)
+		}
+	}
+	return mismatchDetails
+}
+
+// createDeclarativeValidationPanicHandler returns a function with panic recovery logic
+// that will increment metrics and either log or append errors based on feature gate settings.
+func createDeclarativeValidationPanicHandler(errs *field.ErrorList) func() {
+	return func() {
+		if r := recover(); r != nil {
+			// Increment the panic metric counter
+			validationmetrics.Metrics.EmitDeclarativeValidationPanicMetric()
+
+			const errorFmt = "panic during declarative validation: %v"
+			if utilfeature.DefaultFeatureGate.Enabled(features.DeclarativeValidationTakeover) {
+				// If takeover is enabled, output as a validation error as authorative validator panicked and validation should error
+				*errs = append(*errs, field.InternalError(nil, fmt.Errorf(errorFmt, r)))
+			} else {
+				// if takeover not enabled, log the panic as a warning
+				klog.Warningf(errorFmt, r)
+			}
+		}
+	}
+}
+
+// WithRecover wraps a validation function with panic recovery logic.
+// It takes a validation function with the ValidateDeclaratively signature
+// and returns a function with the same signature.
+// The returned function will execute the wrapped function and handle any panics by
+// incrementing the panic metric, and logging an error message
+// if DeclarativeValidationTakeover=disabled, and adding a validation error if enabled.
+func WithRecover(
+	validateFunc func(ctx context.Context, options sets.Set[string], scheme *runtime.Scheme, obj runtime.Object) field.ErrorList,
```

Comment:
This function can be declared generically, eliminating the need for a separate implementation for update.